### PR TITLE
ENH: add mean and triangle thresholding algorithms

### PR DESCRIPTION
**Pull request by sciunto**
## Description

This PR adds two new algorithms:

* mean (trivial)
* triangle (less trivial).

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Unit tests

### CODE REVIEW

File: skimage/filters/thresholding.py [hunk 1]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
             new_thresh = temp + tolerance
 
     return threshold + immin
+
+
+def threshold_mean(image):
+    """Return threshold value based on the of grayscale values.
```
Comment (soupault) [inline skimage/filters/thresholding.py]: Something is missing.

File: skimage/filters/thresholding.py [hunk 2]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+
+    Parameters
+    ----------
+    image : array
```
Comment (soupault) [inline skimage/filters/thresholding.py]: Specify dimensionality.

File: skimage/filters/thresholding.py [hunk 3]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+        Grayscale input image.
+
+    Returns
+    -------
+    threshold : float
+        Upper threshold value. All pixels intensities that less or equal of
```
Comment (soupault) [inline skimage/filters/thresholding.py]: 'Less than or equal to'.

File: skimage/filters/thresholding.py [hunk 4]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_mean(image)
+    >>> binary = image > thresh
+    """
+    return np.mean(image.ravel())
+
+
+def threshold_triangle(image, nbins=256):
+    """Return threshold value based on triangle algorithm.
```
Comment (soupault) [inline skimage/filters/thresholding.py]: ..._the_ triangle...

File: skimage/filters/thresholding.py [hunk 5]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+
+    Parameters
+    ----------
+    image : array
```
Comment (soupault) [inline skimage/filters/thresholding.py]: Same as above.

File: skimage/filters/thresholding.py [hunk 6]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+        Grayscale input image.
+    nbins : int, optional
+        Number of bins used to calculate histogram. This value is ignored for
+        integer arrays.
+
+    Returns
+    -------
+    threshold : float
+        Upper threshold value. All pixels intensities that less or equal of
```
Comment (soupault) [inline skimage/filters/thresholding.py]: Same as above.

File: skimage/filters/thresholding.py [hunk 7]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+    >>> image = camera()
+    >>> thresh = threshold_triangle(image)
+    >>> binary = image > thresh
+    """
+    # nbins is ignored for interger arrays
+    # so, we recalculate the effective nbins.
+    hist, bin_centers = histogram(image.ravel(), nbins)
+    nbins = bin_centers[-1] - bin_centers[0] + 1
+
+    # Find peak and histogram extremities.
```
Comment (soupault) [inline skimage/filters/thresholding.py]: 'extremums'?

File: skimage/filters/thresholding.py [hunk 8]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+    argpeak_height = np.argmax(hist)
+    peak_height = np.max(hist)
```
Comment (soupault) [inline skimage/filters/thresholding.py]: `peak_height = hist[argpeak_height]`?

File: skimage/filters/thresholding.py [hunk 9]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+    arglefth = np.where(hist>0)[0][0]
+    argrighth = np.where(hist>0)[0][-1]
```
Comment (soupault) [inline skimage/filters/thresholding.py]: These variable names are too hard to read. Could you rename them, please or add some underscores?

**Comment by soupault** [comment]
@sciunto Thanks for the contribution! `mean` thresholding will be our killer-feature :trollface: .

What about gallery examples though? Would be nice to have one at least for the `triangle` method (applied to several images we ship).

**Comment by sciunto** [comment]
@soupault Fixed. I really hope that the mean threshold will be printed in gold letters in the next announcement! :)

For the gallery, it's a good point. My opinion is that it's fairly useless to accumulate examples more or less identical to [this one](http://scikit-image.org/docs/0.12.x/auto_examples/segmentation/plot_otsu.html#example-segmentation-plot-otsu-py) with few changes. Also, this will make the gallery page longer and therefore, it will more difficult to find something in there in the future. 

Instead, I think that a more comprehensive page on thresholding will be more useful for users. This is part of the objectives for #2110, but maybe I will do that in an extra PR, because we already have several PRs on thresholding and it's difficult to handle everything in parallel.

File: skimage/filters/thresholding.py [hunk 10]
```diff
@@ -458,37 +458,37 @@ def threshold_triangle(image, nbins=256):
...
     nbins = bin_centers[-1] - bin_centers[0] + 1
 
-    # Find peak and histogram extremities.
-    argpeak_height = np.argmax(hist)
-    peak_height = np.max(hist)
-    arglefth = np.where(hist>0)[0][0]
-    argrighth = np.where(hist>0)[0][-1]
+    # Find peak, lowest and highest gray levels.
+    arg_peak_hight = np.argmax(hist)
+    peak_height = hist[argpeak_height]
```
Comment (soupault) [inline skimage/filters/thresholding.py]: `arg_peak_hight` + `argpeak_height` == `arg_peak_height` :laughing:

Comment (sciunto) [inline skimage/filters/thresholding.py]: Bad regexp :)

**Comment by soupault** [comment]
@sciunto Ah, now I see what are you up to...
I like your idea a lot! `thresholding` section of the documentation could be a very useful and informative thing to have.
So, this way we should have (1) the section mentioned above with the results of _different_ algorithms on _different_ images - so to give user an intuition what kind of thresholding do we have and do he or she needs, (2) `try_all` function to evaluate the algorithms on user images to refine his or her intuition and to check the most suitable algorithm.
By the way, I remember that `sklearn` team has done something very similar for, for example, [clustering](http://scikit-learn.org/stable/modules/clustering.html).

File: skimage/filters/tests/test_thresholding.py [hunk 11]
```diff
@@ -273,5 +277,37 @@ def test_isodata_moon_image_negative_float():
...
+    t_inv_img = inv_img > t
+    t_inv_inv_img = np.invert(t_inv_img)
+
+    t = threshold_triangle(img)
+    t_img = img > t
+
+    # Check that most of the pixels are identical
+    # See numpy #7685 for a future np.testing API
+    unequal_pos = np.where(t_img.ravel() != t_inv_inv_img.ravel())
+    assert(len(unequal_pos[0])/len(t_img.ravel()) < 1e-2)
```
Comment (soupault) [inline skimage/filters/tests/test_thresholding.py]: Whitespaces around `/`

File: skimage/filters/thresholding.py [hunk 12]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+
+    Parameters
+    ----------
+    image : (N, M) ndarray
+        Grayscale input image.
+
+    Returns
+    -------
+    threshold : float
+        Upper threshold value. All pixels intensities more than
```
Comment (soupault) [inline skimage/filters/thresholding.py]: Maybe 'pixels with intensities`?

File: skimage/filters/thresholding.py [hunk 13]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+    hist, bin_centers = histogram(image.ravel(), nbins)
+    nbins = bin_centers[-1] - bin_centers[0] + 1
+
+    # Find peak, lowest and highest gray levels.
+    arg_peak_height = np.argmax(hist)
+    peak_height = hist[arg_peak_height]
+    arg_low_level = np.where(hist>0)[0][0]
+    arg_high_level = np.where(hist>0)[0][-1]
+
+    # Flip is true if left tail is shorter.
```
Comment (soupault) [inline skimage/filters/thresholding.py]: true -> True

File: skimage/filters/thresholding.py [hunk 14]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+    flip = arg_peak_height - arg_low_level < arg_high_level - arg_peak_height
+    if flip:
+        hist = hist[::-1]
+        arg_low_level = nbins - arg_high_level - 1
+        arg_peak_height = nbins - arg_peak_height - 1
+
+    # If flip == True, argright becomes incorrect
```
Comment (soupault) [inline skimage/filters/thresholding.py]: argright -> arg_high_level

File: skimage/filters/thresholding.py [hunk 15]
```diff
@@ -389,3 +391,109 @@ def threshold_li(image):
...
+        algorithms," CVGIP: Graphical Models and Image Processing,
+        vol. 55, pp. 532-537, 1993.
+
+    Examples
+    --------
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_mean(image)
+    >>> binary = image > thresh
+    """
```
Comment (soupault) [inline skimage/filters/thresholding.py]: We need to check here for the correct `image` shape and the number of dimensions.
Same for :triangular_ruler:.

Comment (sciunto) [inline skimage/filters/thresholding.py]: Good point. Actually most of the functions in this file misses this check.

**Comment by sciunto** [comment]
@soupault Thanks for your comments!

Do you agree to replace the check and the warning in `threshold_otsu` by `assert_nD(image, 2)`?

**Comment by soupault** [comment]
@sciunto let's keep `otsu` warning for another PR and discussion. Regarding that particular case you should bother @jni, [Keeper, Guardian and Evangelist of n-dimensions, Honoured Member of the Hilbert Chamber](https://github.com/scikit-image/scikit-image/pull/1876#issuecomment-171853936).

**Comment by sciunto** [comment]
Rebasing this is aweful. I'll open another PR.

**Pull request closed without merging**